src/http/client.cc

```diff
@@ -130,13 +130,30 @@ future<> connection::close() {
     return when_all(_read_buf.close(), _write_buf.close()).discard_result();
 }
 
+class basic_connection_factory : public connection_factory {
+    ipv4_addr _addr;
+public:
+    explicit basic_connection_factory(ipv4_addr addr)
+            : _addr(std::move(addr))
+    {
+    }
+    virtual future<connected_socket> make() override {
+        return seastar::connect(make_ipv4_address(_addr), {}, transport::TCP);
+    }
+};
+
 client::client(ipv4_addr addr)
-        : _addr(std::move(addr))
+        : client(std::make_unique<basic_connection_factory>(std::move(addr)))
+{
+}
+
+client::client(std::unique_ptr<connection_factory> f)
+        : _new_connections(std::move(f))
 {
 }
 
 future<client::connection_ptr> client::get_connection() {
-    return seastar::connect(make_ipv4_address(_addr), {}, transport::TCP).then([] (connected_socket cs) {
+    return _new_connections->make().then([] (connected_socket cs) {
```

Comment:
Should http_client be local to a specific address, or should we get an address each time?

Even if you talk to one endpoint, DNS may resolve it to multiple IP addresses.

Reply:
It's up to the factory currently. The basic_connection_factory gets ~ipv4_~ socket_address and is thus acting the former. To get the latter we could have ... hostname_connection_factory?

Reply:
Since it's a customization point, it's okay. Since it's so common we may in include it in the basic implementation.